### PR TITLE
Allow configuring Android SDK name and type for intellij project modules

### PR DESCRIPTION
**Pull request by kageiit**
Ref #821

This adds support for setting the default android sdk name and type in generated intellij iml files.

```
[intellij]
  android_sdk_name = Android API 23 Platform
  android_sdk_type = AndroidSDK
```

### CODE REVIEW

**Comment by facebook-github-bot-2** [comment]
By analyzing the blame information on this pull request, we identified @alsutton, @ilya-klyuchnikov and @andrewjcg to be potential reviewers.

**Comment by facebook-github-bot-6** [comment]
@kageiit updated the pull request.

File: src/com/facebook/buck/jvm/java/intellij/IjModuleFactory.java [hunk 1]
```diff
@@ -670,6 +683,7 @@ public void apply(TargetNode<JavaTestDescription.Arg> target, ModuleBuildContext
           true /* wantsPackagePrefix */,
           context);
       addCompiledShadowIfNeeded(target, context);
+      context.setLanguageLevel(getLevel(target.getConstructorArg().source));
```
Comment (marcinkosiba) [inline src/com/facebook/buck/jvm/java/intellij/IjModuleFactory.java]: multiple TargetNodes in the same module could have different language levels. The way you've written it means "last one wins", at minimum keep them as a ImmutableSortedSet<String> and get the first/last one.

File: src/com/facebook/buck/jvm/java/intellij/IjProjectWriter.java [hunk 2]
```diff
@@ -109,6 +110,15 @@ private Path writeModule(IjModule module) throws IOException {
         "androidFacet",
         projectDataPreparer.getAndroidProperties(module));
 
+    Optional<String> javaSdk = buckConfig.getValue("intellij", "java_sdk");
```
Comment (marcinkosiba) [inline src/com/facebook/buck/jvm/java/intellij/IjProjectWriter.java]: IjProjectWriter should cary as little logic and behavior as possible as it's only covered by integration tests. Move this to IjProjectTemplateDataPreparer and add tests.

Also, passing in the entire BuckConfig is overkill as, again, it makes things harder to test (I know the IJProjectCleaner does it, however, you will notice that that class has no tests :/). We have two patterns:
 1) create a IjProjectBuckConfig and have it wrap BuckConfig (see `ArtifactCacheBuckConfig` for an example), this has the benefit of keeping the config-ish logic in one place, but to test you still need to new up a BuckConfig,
 2) create a data object for the config and have separate code that extracts it from BuckConfig (see `AbstractRageConfig`, and `RageBuckConfig`).
I'd prefer using 2), as it makes testing simpler, but if you go 1) I won't block the review.

**Comment by marcinkosiba** [comment]
Have you considered what happens when aggregation is enabled? I'd expect modules with different levels to not want to merge with each other.

I'd like to see tests for the following scenarios:
 - how are multiple language versions resolved if 2 targets in the same path have different settings
 - a test showing that the buckconfig value is the fallback
 - a test showing how this interacts with the module aggregation code.

File: src/com/facebook/buck/jvm/java/intellij/ij-module.st [hunk 3]
```diff
@@ -49,7 +49,14 @@
         %if(generatedFolder.isAndroidResources)% type="java-resource" %endif%/>
     </content>
 }%
+
+%if(!androidFacet.enabled && java_sdk)%
+    <orderEntry type="jdk" jdkName="%java_sdk%" jdkType="JavaSDK" />
```
Comment (marcinkosiba) [inline src/com/facebook/buck/jvm/java/intellij/ij-module.st]: I had a bunch of code that would attempt to read IntelliJ's config and determine whether these SDKs are actually defined (with the intent to figure out if the .jar files they include would be the ones buck would use during the build). IIRC that was like 80% complete, would you be interested in taking that over?

Assuming that the developers will have the correct SDK set up has not worked well for us internally. Any changes to the default are painful and it takes a long time for everyone to update their IJ setup.

Comment (kageiit) [inline src/com/facebook/buck/jvm/java/intellij/ij-module.st]: This is purely scoped to java_sdk. If someone wants to use groosdk, it would need to be a separate config key `groovy_sdk` in the intellij block of buckconfig and could be set here.

I would like to get this in to atleast mirror the legacy support and make it module level granularity. The auto detection change should not be part of this change imo

Comment (kageiit) [inline src/com/facebook/buck/jvm/java/intellij/ij-module.st]: > I had a bunch of code that would attempt to read IntelliJ's config

What if there are multiple instalaltions of inteeli/android studio or its variants? There is a lot of complexity involved in that calculation and it should be kept separate from this change

**Comment by kageiit** [comment]
> Have you considered what happens when aggregation is enabled?

I dont know what aggregation means. Could you please clarify?

> how are multiple language versions resolved if 2 targets in the same path have different settings

I am not sure if intellij supports this? Is this a common enough scenario to support?
I want to keep this change as small as possible to support the most common case. More features etc. can be added in later

File: src/com/facebook/buck/jvm/java/intellij/ij-module.st (hunk 3 above)
Comment (marcinkosiba) [inline src/com/facebook/buck/jvm/java/intellij/ij-module.st]: > This is purely scoped to java_sdk. If someone wants to use groovysdk, they would need to be a separate config key groovy_sdk in the intellij block of buckconfig and it could be used here.

I think one of my comments wasn't clear. I didn't mean you should add support for more kind of language SDKs, I meant that you have no guarantee that these SDKs are defined in IntelliJ's config nor that they are correctly set up. To put it another way - you're effectively hardcoding strings and expecting those to be correctly set up on every machine you run `buck project` on.

> The auto detection change should not be part of this change imo
Agreed, that was not what I was asking for. My question was "do you want the code?", not "merge it into this PR". The real question behind it is: are you planning more changes to this feature (project generation)?

File: src/com/facebook/buck/jvm/java/intellij/ij-module.st [hunk 4]
```diff
@@ -49,7 +49,14 @@
         %if(generatedFolder.isAndroidResources)% type="java-resource" %endif%/>
     </content>
 }%
+
+%if(!androidFacet.enabled && java_sdk)%
```
Comment (marcinkosiba) [inline src/com/facebook/buck/jvm/java/intellij/ij-module.st]: just like with IjProjectWriter, the templates are only tested through integration tests, so it's more convenient to have any logic live at higher abstraction levels. Here I suggest adding a Jdk type on the module, if it's null, print 'inheritedJdk', otherwise print it's .name and .type:

```
%if(jdk)%
  <orderEntry type="jdk" jdkName="%jdk.name%" jdkType="%jdk.type%" />
%else%
 <orderEntry type="inheritedJdk" />
%endif%
```

**Comment by marcinkosiba** [comment]
>> Have you considered what happens when aggregation is enabled?

> I dont know what aggregation means. Could you please clarify?

aggregates multiple modules into one, see https://github.com/facebook/buck/blob/master/src/com/facebook/buck/jvm/java/intellij/IjModuleGraph.java

>> how are multiple language versions resolved if 2 targets in the same path have different settings

> I am not sure if intellij supports this? Is this a common enough scenario to support?
> I want to keep this change as small as possible to support the most common case. More features etc. can be added in later

Correct, intellij doesn't support this, but buck does. We need a deterministic way to translate the more expressive buck model into something IntelliJ can understand. Yes, this is common enough to support.

**Comment by kageiit** [comment]
Seems like part of this was addressed in 60058a4f32681808ce5776d9f8bbb348282b5633

I will rebase this pr and include only the android sdk changes

**Comment by facebook-github-bot-7** [comment]
@kageiit updated the pull request - [view changes](https://github.com/facebook/buck/pull/832/files/2d27703f0910b2823ca312bcb9753be11c3f9ed8..b6aa77cebeb690e1185b1ad030abbffca6a5e28d)

**Comment by kageiit** [comment]
@marcinkosiba I have rewritten this with your suggestions and isolated all logic to `IjModuleFactory`. Let me know if this approach looks ok and I will add tests to check all the scenarios.

**Comment by facebook-github-bot-9** [comment]
@kageiit updated the pull request - [view changes](https://github.com/facebook/buck/pull/832/files/b6aa77cebeb690e1185b1ad030abbffca6a5e28d..353a6e532e506053e6e6310ffce5a2516ea42d0b)

**Comment by facebook-github-bot-8** [comment]
@kageiit updated the pull request - [view changes](https://github.com/facebook/buck/pull/832/files/353a6e532e506053e6e6310ffce5a2516ea42d0b..5bea36ce9d7cf5cd2c6393bc2b5753bb2633e5a6)

**Comment by facebook-github-bot-8** [comment]
@kageiit updated the pull request - [view changes](https://github.com/facebook/buck/pull/832/files/5bea36ce9d7cf5cd2c6393bc2b5753bb2633e5a6..6d6adedb847210e9ddb0624c43c35ab2b23e3fb8)

File: src/com/facebook/buck/jvm/java/intellij/IjProjectTemplateDataPreparer.java [hunk 5]
```diff
@@ -400,6 +400,31 @@ public ModuleIndexEntry apply(IjModule module) {
     return androidProperties;
   }
 
+  public Map<String, String> getJdkProperties(IjModule module)  throws IOException {
```
Comment (marcinkosiba) [inline src/com/facebook/buck/jvm/java/intellij/IjProjectTemplateDataPreparer.java]: Why do you to convert this to a map? ST handles Java objects just fine.

Comment (marcinkosiba) [inline src/com/facebook/buck/jvm/java/intellij/IjProjectTemplateDataPreparer.java]: double space after `throws`

File: src/com/facebook/buck/jvm/java/intellij/IjModuleFactory.java [hunk 6]
```diff
@@ -359,29 +361,65 @@ public IjModule createModule(
...
+      chosenJdkType = intellijConfig.getAndroidSdkType().or(DEFAULT_ANDROID_SDK_TYPE);
+    } else {
+      if (chosenJdkName.isPresent()) {
+        String jdkName = chosenJdkName.get();
+        if (!jdkName.contains(".")) {
+          chosenJdkName = Optional.of(String.format("1.%s", jdkName));
+        }
       }
+
+      chosenJdkName = chosenJdkName.or(intellijConfig.getJdkName());
```
Comment (marcinkosiba) [inline src/com/facebook/buck/jvm/java/intellij/IjModuleFactory.java]: this is confusing, put it in an `else` clause of the `if` statement above

Comment (marcinkosiba) [inline src/com/facebook/buck/jvm/java/intellij/IjModuleFactory.java]: We already have a global language and target level configured in buck: https://buckbuild.com/rule/java_library.html (`source`/`target`) and https://buckbuild.com/concept/buckconfig.html#java.source_level

File: src/com/facebook/buck/jvm/java/intellij/IjProjectTemplateDataPreparer.java (hunk 5 above)
Comment (kageiit) [inline src/com/facebook/buck/jvm/java/intellij/IjProjectTemplateDataPreparer.java]: ah I was not aware. I will clean it up :)

File: src/com/facebook/buck/jvm/java/intellij/IjModuleFactory.java (hunk 6 above)
Comment (kageiit) [inline src/com/facebook/buck/jvm/java/intellij/IjModuleFactory.java]: Sure I will make the changes. Thanks for the feedback

**Comment by kageiit** [comment]
Seems like the approach looks fine. So i will address comments from @marcinkosiba and add tests

**Comment by kageiit** [comment]
Just did a quick rebase. will be adding tests tomorrow

File: test/com/facebook/buck/jvm/java/intellij/IjModuleFactoryTest.java [hunk 7]
```diff
@@ -525,50 +525,10 @@ public void testOverrideSdkFromBuckConfig() throws Exception {
         moduleBasePath,
         ImmutableSet.of(java8Node));
 
-    assertThat(moduleWithDefault.getSdkName(), equalTo(Optional.<String>absent()));
-    assertThat(moduleWithJava8.getSdkName(), equalTo(Optional.of("TestSDK")));
-    assertThat(moduleWithJava8.getLanguageLevel(), equalTo(Optional.of("1.8")));
-  }
-
-  @Test
-  public void testAndroidPrebuiltAar() {
```
Comment (marcinkosiba) [inline test/com/facebook/buck/jvm/java/intellij/IjModuleFactoryTest.java]: why did you remove this?

**Comment by kageiit** [comment]
I have not added tests yet. See my previous comment. I just did a quick rebase to resolve some merge conflicts. I will fix the line limit and test scenarios

**Comment by styurin** [comment]
This is already implemented, can you check existing functionality and see if we are missing something?

**Comment by kageiit** [comment]
@styurin please take a look

File: docs/concept/buckconfig.soy [hunk 8]
```diff
@@ -1570,16 +1570,34 @@ $ buck targets --resolve-alias app#src_jar
   {param name: 'jdk_name' /}
   {param example_value: 'Java SDK 1.6' /}
   {param description}
-    IntelliJ project SDK name.
+    IntelliJ project Java SDK name.
```
Comment (styurin) [inline docs/concept/buckconfig.soy]: `jdk_name` is a bad name here, it's an SDK name, not JDK name. It's better leave it as it was before (i.e. just SDK name) because it can point to an Android SDK or Intellij Plugin SDK.

File: docs/concept/buckconfig.soy [hunk 9]
```diff
@@ -1570,16 +1570,34 @@ $ buck targets --resolve-alias app#src_jar
...
 {/call}
 
 {call buckconfig.entry}
   {param section: 'intellij' /}
   {param name: 'jdk_type' /}
-  {param example_value: 'Android SDK or JavaSDK' /}
+  {param example_value: 'JavaSDK' /}
   {param description}
-    IntelliJ project SDK type.
+    IntelliJ project Java SDK type.
```
Comment (styurin) [inline docs/concept/buckconfig.soy]: the same, JDK is one type, there might be other types.

File: docs/concept/buckconfig.soy [hunk 10]
```diff
@@ -1570,16 +1570,34 @@ $ buck targets --resolve-alias app#src_jar
...
+  {/param}
+{/call}
+
+{call buckconfig.entry}
+  {param section: 'intellij' /}
+  {param name: 'android_sdk_type' /}
```
Comment (styurin) [inline docs/concept/buckconfig.soy]: We already have `java_library_sdk_names`, can you follow the same logic and rename this option to `android_library_sdk_type`?

File: docs/concept/buckconfig.soy [hunk 11]
```diff
@@ -1570,16 +1570,34 @@ $ buck targets --resolve-alias app#src_jar
...
+  {param example_value: 'Android SDK' /}
+  {param description}
+    IntelliJ project Android SDK type.
```
Comment (styurin) [inline docs/concept/buckconfig.soy]: "Default SDK type for android_library modules" or something like that. In other words, please, indicate that this is applied to android_library only.

File: docs/concept/buckconfig.soy [hunk 12]
```diff
@@ -1570,16 +1570,34 @@ $ buck targets --resolve-alias app#src_jar
...
+  {/param}
+{/call}
+
+{call buckconfig.entry}
+  {param section: 'intellij' /}
+  {param name: 'android_sdk_name' /}
```
Comment (styurin) [inline docs/concept/buckconfig.soy]: The same, `android_library_sdk_name`.

File: docs/concept/buckconfig.soy [hunk 13]
```diff
@@ -1570,16 +1570,34 @@ $ buck targets --resolve-alias app#src_jar
...
+  {param example_value: 'Google Inc.:Google APIs:24' /}
+  {param description}
+    IntelliJ project Android SDK name.
```
Comment (styurin) [inline docs/concept/buckconfig.soy]: The same comment as in SDK type

File: src/com/facebook/buck/jvm/java/intellij/AbstractIjProjectConfig.java [hunk 14]
```diff
@@ -48,6 +48,10 @@ public boolean isAutogenerateAndroidFacetSourcesEnabled() {
 
   public abstract Optional<String> getProjectJdkType();
 
+  public abstract Optional<String> getProjectAndroidSdkName();
```
Comment (styurin) [inline src/com/facebook/buck/jvm/java/intellij/AbstractIjProjectConfig.java]: `Project` means it's a project level setting. In this case methods should be `getAndroidLibrarySdkName` and `getAndroidLibrarySdkType`.

File: docs/concept/buckconfig.soy [hunk 15]
```diff
@@ -1570,16 +1570,34 @@ $ buck targets --resolve-alias app#src_jar
...
+  {param example_value: 'Android SDK' /}
+  {param description}
+    IntelliJ project Android SDK type.
+  {/param}
+{/call}
+
+{call buckconfig.entry}
+  {param section: 'intellij' /}
+  {param name: 'android_sdk_name' /}
+  {param example_value: 'Google Inc.:Google APIs:24' /}
```
Comment (styurin) [inline docs/concept/buckconfig.soy]: Can you replace the example with the default SDK name, for example, "Android API 23 Platform"

File: test/com/facebook/buck/jvm/java/intellij/testdata/experimental_project_with_custom_sdks/.buckconfig [hunk 16]
```diff
@@ -9,3 +9,5 @@
 [intellij]
   java_library_sdk_names = 1.6 => Java SDK 1.6, \
                            1.8 => Java SDK 1.8
+  android_sdk_name = "Google Inc.:Google APIs:24"
```
Comment (styurin) [inline test/com/facebook/buck/jvm/java/intellij/testdata/experimental_project_with_custom_sdks/.buckconfig]: The same about sdk name, please, replace with the default name "Android API 23 Platform".

File: test/com/facebook/buck/jvm/java/intellij/testdata/experimental_project_with_custom_sdks/modules/libA/src/main/AndroidManifest.xml [hunk 17]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (styurin) [inline test/com/facebook/buck/jvm/java/intellij/testdata/experimental_project_with_custom_sdks/modules/libA/src/main/AndroidManifest.xml]: Do we really need real code here?

File: docs/concept/buckconfig.soy (hunk 10 above)
Comment (kageiit) [inline docs/concept/buckconfig.soy]: This applies to all android rules. I think the original argument should instead be renamed to `java_sdk_names` instead because it would apply to all java rules as well

File: docs/concept/buckconfig.soy (hunk 12 above)
Comment (kageiit) [inline docs/concept/buckconfig.soy]: Same comment as above. I think this should be named `android_sdk_name` as it applies to all android rules

File: test/com/facebook/buck/jvm/java/intellij/testdata/experimental_project_with_custom_sdks/modules/libA/src/main/AndroidManifest.xml (hunk 17 above)
Comment (kageiit) [inline test/com/facebook/buck/jvm/java/intellij/testdata/experimental_project_with_custom_sdks/modules/libA/src/main/AndroidManifest.xml]: We dont. I will remove
